server/control.go

```diff
@@ -454,9 +460,18 @@ func (ctl *Control) manager() {
 				ctl.CloseProxy(m)
 				xl.Info("close proxy [%s] success", m.ProxyName)
 			case *msg.Ping:
+				if err := ctl.authVerifier.VerifyPing(m); err != nil {
+					xl.Warn("received invalid ping: %v", err)
+					ctl.sendCh <- &msg.Pong{
+						Error: "invalid authentication in ping",
+					}
+					return
+				}
 				ctl.lastPing = time.Now()
 				xl.Debug("receive heartbeat")
-				ctl.sendCh <- &msg.Pong{}
+				ctl.sendCh <- &msg.Pong{
+					Error: "",
```

Comment:
it defaults to an empty string.

Reply:
Yeah true, I thought it would be better explicitly but maybe it's a waste

Reply:
In general I favor explicit over implicit too. In this case however, making the developer aware of the attribute Error here does not really add much. For me this is non-voting so as you wish :)